### PR TITLE
Add Final fees to PricelessPositionManager

### PR DESCRIPTION
**Pull request by nicholaspai**
One half of #876, the other being adding final fees to `Liquidatable`

Pays the final fee on `expire()` when `_requestOraclePrice` is called

___
I made one subtle design decision:

- When paying final fees, I decided to spread fees across `totalPositionCollateral` instead of `pfc`. If PfC !== locked-collateral, then it is possible that liquidations are not included in the PfC number. I don't think that fees should include assets being liquidated (since it is possible that liquidations do not trigger `requestOraclePrice` and any disputes will pay their own separate final fees):

```
FixedPoint.Unsigned memory feePercentage = totalPaid.div(totalPositionCollateral);
// Compute adjustment to be applied to the position collateral (1 - feePercentage).
FixedPoint.Unsigned memory adjustment = FixedPoint.fromUnscaledUint(1).sub(feePercentage);
``` 
___
Also, I ran into a rounding bug and had to force the amount of collateral and synthetic to work without rounding. Here is what happened for documentation purposes:

- When writing the last test in `PricelessPositionManager.js`, I initially set up the test with `150 collateral`, `100 synthetic`, and `final fee = 1`. After calling `expire()` I was expecting `getCollateral()` to return `149000000000000000000` but it was returning `149000000000000000100`.This is likely coming from the `feeAdjustment` modification code that I included above, specifically the `div()`

### CODE REVIEW

File: core/contracts/financial_templates/FeePayer.sol [hunk 1]
```diff
@@ -70,6 +70,22 @@ contract FeePayer is Testable {
         return regularFee.add(latePenalty);
     }
 
+    /**
+     * @notice Pays UMA DVM final fees to the Store contract. This is a flat fee charged for each price request.
+     * @return the amount of collateral that was paid to the Store.
+     */
+    function payFinalFees() public returns (FixedPoint.Unsigned memory totalPaid) {
+        StoreInterface store = StoreInterface(finder.getImplementationAddress("Store"));
```
Comment (mrice32) [inline core/contracts/financial_templates/FeePayer.sol]: nit: you can merge master and use the _getStore() method:
```
StoreInterface store = _getStore();
```

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 2]
```diff
@@ -451,7 +451,31 @@ contract PricelessPositionManager is FeePayer {
         oracle.requestPrice(priceIdentifer, requestedTime);
     }
 
-    function _getOraclePrice(uint requestedTime) internal view returns (FixedPoint.Unsigned memory) {
+    function _payFinalFee() internal {
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: nit: maybe make this private if you don't want subclasses to use it to pay their final fees (in the case of dispute, for instance)?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 3]
```diff
@@ -453,7 +453,31 @@ contract PricelessPositionManager is FeePayer {
...
+        // Send the fee payment.
+        FixedPoint.Unsigned memory totalPaid = super.payFinalFees();
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Why are you using `super` here? Why not just call `payFinalFees()`? I'm assuming you're not trying to pass the call up the inheritance tree -- you just want to call the most most derived/virtual `payFinalFees()` function, right?

File: core/contracts/financial_templates/FeePayer.sol [hunk 4]
```diff
@@ -70,6 +70,22 @@ contract FeePayer is Testable {
         return regularFee.add(latePenalty);
     }
 
+    /**
+     * @notice Pays UMA DVM final fees to the Store contract. This is a flat fee charged for each price request.
+     * @return the amount of collateral that was paid to the Store.
+     */
+    function payFinalFees() public returns (FixedPoint.Unsigned memory totalPaid) {
```
Comment (mrice32) [inline core/contracts/financial_templates/FeePayer.sol]: Is there any reason this should be public? Couldn't people drain the contract by repeatedly calling this method from the outside? By my understanding it should only be callable from inside the contract, right?

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 5]
```diff
@@ -453,7 +453,31 @@ contract PricelessPositionManager is FeePayer {
...
+
+        // Compute adjustment to be applied to the position collateral (1 - feePercentage).
+        FixedPoint.Unsigned memory adjustment = FixedPoint.fromUnscaledUint(1).sub(feePercentage);
+
+        // Apply fee percentage to adjust totalPositionCollateral and positionFeeAdjustment.
+        totalPositionCollateral = totalPositionCollateral.mul(adjustment);
+        positionFeeAdjustment = positionFeeAdjustment.mul(adjustment);
+    }
+
+    function _getOraclePrice(uint requestedTime) public view returns (FixedPoint.Unsigned memory) {
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Did you mean to change this to `public`?

File: core/contracts/financial_templates/FeePayer.sol (hunk 1 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/FeePayer.sol]: nice

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 2 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: good point since `Liquidatable` shouldn't be able to call this to pay fin fees

File: core/contracts/financial_templates/FeePayer.sol (hunk 4 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/FeePayer.sol]: ah this was just a typo in copying the code initially from `PricelessPositionManager.payFees()`

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 5 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: ah nope good catch

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 3 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: ah this was just a typo in copying the code initially from `PricelessPositionManager.payFees()`

File: core/contracts/financial_templates/FeePayer.sol (hunk 4 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/FeePayer.sol]: yes good point

Comment (nicholaspai) [reply core/contracts/financial_templates/FeePayer.sol]: this is now `internal`

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 2 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: this is now (back to)` internal`

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 5 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: this is now (back to) `internal`

File: core/contracts/financial_templates/FeePayer.sol [hunk 6]
```diff
@@ -70,6 +70,22 @@ contract FeePayer is Testable {
         return regularFee.add(latePenalty);
     }
 
+    /**
+     * @notice Pays UMA DVM final fees to the Store contract. This is a flat fee charged for each price request.
+     * @return the amount of collateral that was paid to the Store.
+     */
+    function payFinalFees() internal returns (FixedPoint.Unsigned memory totalPaid) {
```
Comment (mrice32) [inline core/contracts/financial_templates/FeePayer.sol]: nits: 
1. You may want to change the name of this function or the _payFinalFee() in the derived class so we don't get them confused.
2. Generally, we use a leading underscore for internal method names.

Comment (nicholaspai) [reply core/contracts/financial_templates/FeePayer.sol]: should I `super` `_payFinalFees()`?

Comment (mrice32) [reply core/contracts/financial_templates/FeePayer.sol]: I wouldn't. I think `super` should really only be used if you're trying to pass control up the inheritance tree iteratively. In this case, you just want to call the `_payFinalFees()` method on the FeePayer contract.

If you name them the same thing you'll be overriding that function, which would be confusing and not really necessary for this case. Note: it was useful for regular fees since the payFees() function can be called externally, and we wanted derived contracts to capture that call. Also, if you override the _payFinalFees() function, you'll also need to make sure you match the return types -- right now, I think they differ.

Comment (nicholaspai) [reply core/contracts/financial_templates/FeePayer.sol]: FYI I did try `super` and all tests passed, but I do agree why `super` wouldn't make sense here. Changing the name now

File: core/contracts/financial_templates/FeePayer.sol [hunk 7]
```diff
@@ -70,6 +70,25 @@ contract FeePayer is Testable {
...
+     * @notice Pays UMA DVM final fees to the Store contract. This is a flat fee charged for each price request.
+     * @return the amount of collateral that was paid to the Store.
+     */
+    function _payFinalFees(address payer) internal returns (FixedPoint.Unsigned memory totalPaid) {
+        StoreInterface store = _getStore();
+        FixedPoint.Unsigned memory finalFee = store.computeFinalFee(address(collateralCurrency));
+
+        if (finalFee.isGreaterThan(0)) {
+            if (payer != address(this)) {
+                require(collateralCurrency.transferFrom(payer, address(this), finalFee.rawValue));
```
Comment (mrice32) [inline core/contracts/financial_templates/FeePayer.sol]: nit: prefer to use the open zeppelin safe ERC20 methods:
```
collateralCurrency.safeTransferFrom(payer, address(this), finalFee.rawValue);
```

File: core/contracts/financial_templates/PricelessPositionManager.sol [hunk 8]
```diff
@@ -453,6 +453,30 @@ contract PricelessPositionManager is FeePayer {
         oracle.requestPrice(priceIdentifer, requestedTime);
     }
 
+    function _payOracleRequestFees() private {
+        // Send the fee payment.
+        FixedPoint.Unsigned memory totalPaid = _payFinalFees(address(this));
+
+        // Exit early if totalCollaterall == 0 to prevent divide by 0.
+        if (totalPositionCollateral.isEqual(FixedPoint.fromUnscaledUint(0))) {
```
Comment (mrice32) [inline core/contracts/financial_templates/PricelessPositionManager.sol]: Instead of checking whether the totalPositionCollateral == 0, shouldn't we make sure that it's greater than the fees (and revert if it's not)? Otherwise, there's not enough collateral in the live positions to pay for the final fee, so the collateral is effectively being pulled from the liquidation pool, meaning someone may not be able to withdraw their liquidation reward.

File: core/test/financial_templates/PricelessPositionManager.js [hunk 9]
```diff
@@ -591,5 +591,112 @@ contract("PricelessPositionManager", function(accounts) {
...
+    const sponsorFinalSynthetic = await tokenCurrency.balanceOf(sponsor);
+
+    // The token Sponsor should gain the value of their synthetics in underlying
+    // + their excess collateral from the over collateralization in their position
+    const expectedSponsorCollateralUnderlying = toBN(amountCollateral)
+      .sub(toBN(numTokens).muln(redemptionPrice))
+      .sub(toBN(finalFeePaid));
+    const expectedSponsorCollateralSynthetic = sponsorInitialSynthetic.muln(redemptionPrice);
+    const totalSponsorCollateralReturned = expectedSponsorCollateralUnderlying + expectedSponsorCollateralSynthetic;
+    assert(sponsorFinalCollateral.sub(sponsorInitialCollateral), totalSponsorCollateralReturned);
```
Comment (mrice32) [inline core/test/financial_templates/PricelessPositionManager.js]: nit: why `assert` vs `assert.equal`?

Here and below.

File: core/contracts/financial_templates/FeePayer.sol (hunk 7 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/FeePayer.sol]: agreed good catch

File: core/contracts/financial_templates/PricelessPositionManager.sol (hunk 8 above)
Comment (nicholaspai) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: yes I think so, but shouldn't we do the same for `payFees`?

```
// Exit early if pfc == 0 to prevent divide by 0.
        if (initialPfc.isEqual(FixedPoint.fromUnscaledUint(0))) {
            return totalPaid;
        }
```

Comment (nicholaspai) [reply core/contracts/financial_templates/PricelessPositionManager.sol]: As discussed in IRL, since final fee is a flat amount and not a %, its possible for `finalFee >= totalPositionCollateral` which will either revert `expire` or render it useless. I replaced the `if` statement above with `require(totalPositionCollateral.isGreaterThan(totalPaid))`

This does raise an edge case in which collateral can be locked in the contract, which I've brought up in #925

File: core/test/financial_templates/PricelessPositionManager.js (hunk 9 above)
Comment (nicholaspai) [reply core/test/financial_templates/PricelessPositionManager.js]: no reason, will modify. In general do we prefer to use assert.isEqual versus assert?

Comment (nicholaspai) [reply core/test/financial_templates/PricelessPositionManager.js]: ah, I did this because `assert` is used instead of `assert.equal` in the "Redemption Post expiry" test. I'll edit those tests too

Comment (nicholaspai) [reply core/test/financial_templates/PricelessPositionManager.js]: yeah weirdly this test does not pass if you replace with `assert.equal` because it tries to compare bignumbers. I tried to `.eq` the two arguments but it doesn't return the exact same big number

Comment (nicholaspai) [reply core/test/financial_templates/PricelessPositionManager.js]: I believe it is a rounding error that is somehow ignored with `assert` these are the two arguments being compared when printed out individually `toString()`:
- `tokenHolderFinalCollateral.sub(tokenHolderInitialCollateral).toString() = 60000000000000000000`
- `expectedTokenHolderFinalCollateral.toString() = 59998198560095364710`

Comment (mrice32) [reply core/test/financial_templates/PricelessPositionManager.js]: The only time we normally use `assert` is when we are testing `didContractThrow`, and that's only because it returns something *truthy*, but not true or false.

Comment (mrice32) [reply core/test/financial_templates/PricelessPositionManager.js]: If there's a rounding error, we should make that part of what we test for and understand why it's there.

Comment (nicholaspai) [reply core/test/financial_templates/PricelessPositionManager.js]: ok I resolved this, it was an error with the `bn.js` `muln` method that doesn't correctly multiply `50000000000000000000` * `1.2`, so i hardcoded the expected answers instead

File: core/test/financial_templates/PricelessPositionManager.js [hunk 10]
```diff
@@ -593,5 +601,141 @@ contract("PricelessPositionManager", function(accounts) {
...
+    await tokenCurrency.approve(pricelessPositionManager.address, sponsorInitialSynthetic, {
+      from: sponsor
+    });
+    await pricelessPositionManager.settleExpired({ from: sponsor });
+    const sponsorFinalCollateral = await collateral.balanceOf(sponsor);
+    const sponsorFinalSynthetic = await tokenCurrency.balanceOf(sponsor);
+
+    // The token Sponsor should gain the value of their synthetics in underlying
+    // + their excess collateral from the over collateralization in their position
+    // Excess collateral = 100 - 50 * 1.2 - 1 = 39
```
Comment (mrice32) [inline core/test/financial_templates/PricelessPositionManager.js]: +1 I really like putting these simple equations in the comments. IMO they make the tests way more readable.

File: core/test/financial_templates/PricelessPositionManager.js [hunk 11]
```diff
@@ -593,5 +601,141 @@ contract("PricelessPositionManager", function(accounts) {
...
+    const expectedSponsorCollateralUnderlying = toBN(toWei("39"));
+    // Value of remaining synthetic tokens = 25 * 1.2 = 30
+    const expectedSponsorCollateralSynthetic = toBN(toWei("30"));
+    const expectedTotalSponsorCollateralReturned = expectedSponsorCollateralUnderlying.add(
+      expectedSponsorCollateralSynthetic
+    );
+    assert(sponsorFinalCollateral.sub(sponsorInitialCollateral), expectedTotalSponsorCollateralReturned);
```
Comment (mrice32) [inline core/test/financial_templates/PricelessPositionManager.js]: Like the other spot, we should probably be using `assert.equal()` here.

Comment (mrice32) [reply core/test/financial_templates/PricelessPositionManager.js]: Just to resolve our earlier conversation, the syntax for assert is:
```
assert(expression, message);
```

So we're effectively just passing expectedTotalSponsorCollateralReturned as the error message, not the thing to compare it to.

Comment (nicholaspai) [reply core/test/financial_templates/PricelessPositionManager.js]: ah i missed this one
